### PR TITLE
[JUJU-4164] Revisit wait-for command errors

### PR DESCRIPTION
**Pull request by SimonRichardson**
The following PR attempts to improve the error messages and consistency
of the wait-for command. The wait-for command was originally intended for
use by the integration test suite and be delivered as part of the plugin 
architecture that juju offers. The wait-for command was upgraded to a 
first-class citizen last year but was never revisited to improve the error 
messages.

A reminder, that as long a query in the wait-for command returns anything
other than falsy, a query is determined as successful. It's all based on boolean
logic. So if `true && true` is fine, but `true && false` is not. The latter will wait
until the timeout is reached. To build up a larger query, it's possible to do the
following o a model with `juju deploy ubuntu -n 3`:

```sh
juju wait-for model --query="forEach(applications, app => forEach(app.units, unit => forEach(unit.machines, machine => int(machine.id) < 4)))" default
```

This states that for all machine ids for a given application is below 4, then
return true. This satisfies the boolean logic. The following would fail given the
same setup:

```sh
juju wait-for model --query="forEach(applications, app => forEach(app.units, unit => forEach(unit.machines, machine => int(machine.id) >= 4)))" default
```

-----

I had planned to improve them as time went on, but never got around to a lot
of the shortcomings. The following aims to start improving the error
messages for all the commands by explaining what went wrong, but more
importantly, how to help fix it. There are a lot of error messages, but the
most common ones have been updated.

In addition, it's now possible to correctly specify a model in the model command
by using the same `<controller>:<model>` or `<model>` syntax as other commands
in juju. Other commands can still use `-m`, it's just that the model command only
accepts the model as an argument.

New functions have also been added, this gives an example of how flexible the
addition of new functions can be. New functions can be added as we go, so PRs
are accepted with good suggestions.

Lastly, the nesting of `forEach` loops is fixed, this was an oversight, but should
at least works correctly now (see QA steps).

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made

## QA steps

```sh
$ juju bootstrap lxd test --build-agent
$ juju add-model default
$ juju deploy ubuntu -n 3
```

### Model not found

```sh
$ juju wait-for model blah
model "blah4" not found, waiting...
```

Then in another process

```sh
$ juju add-model blah
```

This should then return the following from the first process.

```sh
model "blah" is running
properties:
  life: alive
  status: available
```

### Not a valid model name

```sh
$ juju wait-for model blah!
ERROR "blah!" is not valid model name
```

### Target controller using model name

```sh
$ juju wait-for model test:default
```

### Quotation

```sh
$ juju wait-for model --query='"life"' default
$ juju wait-for model --query="'life'" default
$ juju wait-for model --query="\"life\"" default
```

### Quotation error

```sh
juju wait-for model --query="\"life" default
ERROR Cannot parse query: string is not correctly terminated.

1 | "life
    ^
Try adding a closing quote to the string.
```

### Wrong identifier error

```sh
$ juju wait-for model --query="unit" default
ERROR Cannot execute query: invalid identifier.

1 | unit

No possible matches for "unit" on ModelInfo:
Did you mean:

    units

Other possible matches:

    - applications
    - controller-uuid
    - is-controller
    - life
    - machines
    - model-uuid
    - name
    - owner
```

### Invalid AND operator (&&)

```sh
$ juju wait-for model --query="1 && true" default
ERROR Cannot parse query: invalid AND (&&) operator.

1 | 1 && true
    ^
Ensure that each side of the && operator can be evaluated to a boolean.
```

### Invalid OR operator (||)

```sh
$ juju wait-for model --query="1 || true" default
ERROR Cannot parse query: invalid OR (||) operator.

1 | 1 || true
    ^
Ensure that each side of the || operator can be evaluated to a boolean.
```

### Len of machines

```sh
$ juju wait-for model --query="len(machines) > 3" default
$ juju add-machine
```

### Unknow function

```sh
$ juju wait-for model --query="no(units)" default
ERROR Cannot execute query: runtime syntax error.

1 | no(units)

wait-for has no function "no".
Did you mean:

    int

Other possible matches:

    - endsWith
    - forEach
    - len
    - print
    - startsWith
```

#### Models

```sh
$ juju wait-for model --query="forEach(applications, app => forEach(app.units, unit => forEach(unit.machines, machine => int(machine.id) < 4)))" default
```

#### Applications

```sh
$ juju wait-for application --query="forEach(units, unit => print(unit))" ubuntu
```

#### Units

```sh
$ juju wait-for unit --query="forEach(machines, machine => print(machine))" ubuntu/0
```

#### Whitespace doesn't matter

```sh
juju wait-for model --query="
forEach(
    applications,
    app => forEach(
        app.units,
        unit => forEach(
            unit.machines,
            machine => int(machine.id) < 4
        )
    )
)" default
```

## Bug reference

https://bugs.launchpad.net/juju/+bug/2019516

### CODE REVIEW

**Comment by hpidcock** [comment]
What am I waiting for?

**Comment by hmlanigan** [comment]
```
$ juju wait-for model  default
Model "default" is running
properties:
  life: alive
  status: available
```
Which has no units displayed, however, asking for units displays them?
```
$ juju wait-for model  default --query=units
Model "default" is running
properties: {}
units:
  ubuntu/0: {}
  ubuntu/1: {}
  ubuntu/2: {}
```

**Comment by hmlanigan** [comment]
Should the ctx line be printed?, only seen with this query so far.
```
juju wait-for unit --query="forEach(machines, machine => print(machine))" ubuntu/0
{ctx:{idents:map[] children:map[applications:map[] machines:map[] units:map[]]} MachineInfo:0xc000501ba0}
Unit "ubuntu/0" is running
properties:
  machines: {}
machines:
  "0": {}
````

File: cmd/juju/waitfor/application.go
Comment (hmlanigan) [inline cmd/juju/waitfor/application.go:190]: The capitalization of the found and not found Infof messages is inconsistent.  Marking here, but happens with application, unit, model etc. 

```
$ juju wait-for model --query="len(machines) > 3" default
model "default" found, waiting...
Model "default" is running
```

**Comment by hmlanigan** [comment]
I created a model: six, and ran `juju deploy ubuntu --base ubuntu@22.04 jammy`. Without switching models:
```
$ juju wait-for application --query="forEach(units, unit => print(unit))" ubuntu -m six
application "ubuntu" not found, waiting...
application "ubuntu" not found, waiting...
application "ubuntu" not found, waiting...
application "ubuntu" not found, waiting...
application "ubuntu" not found, waiting...
.... 
``` 
Have seen the "not found, waiting" printed more than once before. Though it did stop at 5 times.

**Comment by hmlanigan** [comment]
```
$ juju wait-for application --query="forEach(units, unit => print(unit))" jammy -m six
{ctx:{idents:map[] children:map[applications:map[] machines:map[] units:map[]]} UnitInfo:0xc000002a80 MachineInfos:map[0:0xc00058ba00]}
Application "jammy" is running
properties:
  units: {}
units:
  jammy/0: {}
```
The output is a bit redundant, "properties" has empty units, then "units" is filled out. This is seen many times, but not all. Not seen with `juju wait-for model  default --query=units`

**Review by hmlanigan — COMMENTED** [review]
Initial go over, will look more in depth tomorrow.

**Comment by SimonRichardson** [comment]
> Have seen the "not found, waiting" printed more than once before. Though it did stop at 5 times.

This is just waiting for the application to show up. Should we keep it, or is it not of any use?

**Comment by SimonRichardson** [comment]
> Should the ctx line be printed?, only seen with this query so far.
> 
> ```
> juju wait-for unit --query="forEach(machines, machine => print(machine))" ubuntu/0
> {ctx:{idents:map[] children:map[applications:map[] machines:map[] units:map[]]} MachineInfo:0xc000501ba0}
> Unit "ubuntu/0" is running
> properties:
>   machines: {}
> machines:
>   "0": {}
> ```
> 
> And `juju wait-for application --query="forEach(units, unit => print(unit))" ubuntu`

I think print is a debug method, it's only useful for debugging. Maybe we don't even offer it?

File: cmd/juju/waitfor/application.go
Comment (SimonRichardson) [reply cmd/juju/waitfor/application.go:190]: Good shout.

**Comment by SimonRichardson** [comment]
> Which has no units displayed, however, asking for units displays them? It's unintuitive. How do you know what can be queried for where?

So depending on what you query for, will depend on the results on the output. So if you do not reference the units then you'll not see units. This becomes useful for large model commands, where we might not want to output the whole status in the result. A case-in-point:

```
juju wait-for model --query="life=='alive' && status=='available' && forEach(applications, app => app.status=='active') && forEach(units, unit => unit.agent-status=='idle' && unit.workload-status=='active')" default
Model "default" is running
properties:
  life: alive
  status: available
applications:
  ubuntu:
    status: active
units:
  ubuntu/0:
    agent-status: idle
    workload-status: active
  ubuntu/1:
    agent-status: idle
    workload-status: active
  ubuntu/2:
    agent-status: idle
    workload-status: active
```

Just querying the information will show you what you requested for:

```
$ juju wait-for model --query="life=='alive' && status=='available' && forEach(applications, app => app.status=='active') && forEach(units, unit => unit.agent-status=='idle')" default
Model "default" is running
properties:
  life: alive
  status: available
applications:
  ubuntu:
    status: active
units:
  ubuntu/0:
    agent-status: idle
  ubuntu/1:
    agent-status: idle
  ubuntu/2:
    agent-status: idle
```

In terms of what's available, that should come in a follow-up PR about documentation. This just fixes the errors to make it more useful before documentation lands.

**Comment by hmlanigan** [comment]
Still printing, I say move it to the --verbose flag.
```
{ctx:{idents:map[] children:map[applications:map[] machines:map[] units:map[]]} MachineInfo:0xc000103380}
```

**Comment by hmlanigan** [comment]
> This is just waiting for the application to show up. Should we keep it, or is it not of any use?

With more experience using the feature, its handy

File: cmd/juju/waitfor/waitfor.go
Comment (hmlanigan) [inline cmd/juju/waitfor/waitfor.go:25]: It'd be handy to have a description for the example of what it does

**Review by hmlanigan — APPROVED** [review]
love the wait for on model! LGTM

File: cmd/juju/waitfor/waitfor.go
Comment (SimonRichardson) [reply cmd/juju/waitfor/waitfor.go:25]: Yeah, I've got a whole ticket to now go and fill in the documentation.

**Pull request merged**
